plugins/coordinator/plugin.go

```diff
@@ -99,42 +118,68 @@ func initCoordinator(bootstrap bool, startIndex uint32) (*coordinator.Coordinato
 		return nil, err
 	}
 
-	// initialize the selector
-	selector.Reset()
-
 	return coo, nil
 }
 
 func run(plugin *node.Plugin) {
-	// pass all new solid bundles to the selector
-	onBundleSolid := events.NewClosure(func(cachedBundle *tangle.CachedBundle) {
-		cachedBundle.ConsumeBundle(func(bndl *tangle.Bundle) { // bundle -1
 
-			if bndl.IsInvalidPastCone() || !bndl.IsValid() || !bndl.ValidStrictSemantics() {
-				// ignore invalid bundles or semantically invalid bundles or bundles with invalid past cone
-				return
+	// create a background worker that signals to issue new milestones
+	daemon.BackgroundWorker("Coordinator[MilestoneTicker]", func(shutdownSignal <-chan struct{}) {
+
+		timeutil.Ticker(func() {
+			// issue next milestone
+			select {
+			case nextMilestoneSignal <- struct{}{}:
+			default:
+				// do not block if already another signal is waiting
 			}
+		}, coo.GetInterval(), shutdownSignal)
 
-			selector.OnNewSolidBundle(bndl)
-		})
-	})
+	}, shutdown.PriorityCoordinator)
 
 	// create a background worker that issues milestones
 	daemon.BackgroundWorker("Coordinator", func(shutdownSignal <-chan struct{}) {
-		tanglePlugin.Events.BundleSolid.Attach(onBundleSolid)
-		defer tanglePlugin.Events.BundleSolid.Detach(onBundleSolid)
+		attachEvents()
 
-		// TODO: add some random jitter to make the ms intervals not predictable
-		timeutil.Ticker(func() {
-			err, criticalErr := coo.IssueNextCheckpointOrMilestone()
-			if criticalErr != nil {
-				log.Panic(criticalErr)
-			}
-			if err != nil {
-				log.Warn(err)
+		// bootstrap the network if not done yet
+		if criticalErr := coo.Bootstrap(); criticalErr != nil {
+			log.Panic(criticalErr)
+		}
+
+	coordinatorLoop:
+		for {
+			select {
+			case <-nextCheckpointSignal:
+				// check the thresholds again, because a new milestone could have been issued in the meantime
+				if tipCount, approveeCount := selector.GetStats(); (tipCount < maxTipsCount) && (approveeCount < maxApproveesCount) {
+					continue
+				}
+
+				// issue a checkpoint
+				if err := coo.IssueCheckpoint(); err != nil {
```

Comment:
Can we still integrate the "traditional" checkpoints to limit attackers with a very high throughput? This could still be a helpful mechanic.

Reply:
So reintroduce the root transaction thing? Or what do you mean exactly? The whole reason to remove it was because of bad confirmation rates in my simulation.